### PR TITLE
Support SQLAlchemy v2

### PR DESCRIPTION
**Pull request by hab6**
**Purpose:** Support SQLAlchemy version 2 and maintain backward compatibility with SQLalchemy version 1.

**Testing:** Tested with SQLAlchemy releases [1.4.46](https://github.com/sqlalchemy/sqlalchemy/releases/tag/rel_1_4_46) and [2.0.3](https://github.com/sqlalchemy/sqlalchemy/releases/tag/rel_2_0_3).

**Reference Issue:** https://github.com/ActianCorp/ingres_sa_dialect/issues/5

**Useful Details:**  See [documentation](https://docs.sqlalchemy.org/en/20/core/internals.html#sqlalchemy.engine.default.DefaultDialect.import_dbapi) and [code comment](https://github.com/zzzeek/sqlalchemy/blob/main/lib/sqlalchemy/engine/create.py#L591) for information about migration from `Dialect.dbapi()` to `Dialect.import_dbapi()`.

### CODE REVIEW

File: lib/ingres_sa_dialect/base.py [hunk 1]
```diff
@@ -686,12 +688,13 @@ def has_sequence(self, connection, sequence_name, schema=None):
                 rs.close()
     
     def get_default_schema_name(self, connection):
-        sqltext = """SELECT dbmsinfo('username')"""
-        
         rs = None
         try:
-            # TODO consider using exec_driver_sql() instead
-            rs = connection.execute(sqltext)
+            if (int((sqlalchemy.__version__).split('.')[0]) >= 2):
```
Comment (clach04) [inline lib/ingres_sa_dialect/base.py:693]: great! consider:

    tuple(map(int, __version__.split('.'))) > (0, 1)

I can't recall off the top of my head if this works with lists. Definitely works with tuples

File: lib/ingres_sa_dialect/ingresdbi.py [hunk 2]
```diff
@@ -8,6 +8,7 @@
 http://esd.ingres.com.
 """
 from ingres_sa_dialect.base import IngresDialect
```
Comment (clach04) [inline lib/ingres_sa_dialect/ingresdbi.py:10]: not part of this change but best-practice to move our import below third party imports - do at the same time as this change while here.

See https://peps.python.org/pep-0008/#imports

Semi-related check out:

  * black (very opinionated code formatter)
  * isort (I've only used this a few times in anger, I was very impressed with it)
  * pyflakes (or some other linting tool)

File: lib/ingres_sa_dialect/ingresdbi.py [hunk 3]
```diff
@@ -17,9 +18,14 @@ class Ingres_ingresdbi(IngresDialect):
     def __init__(self, **kwargs):
         IngresDialect.__init__(self, **kwargs)
 
-    @classmethod
-    def dbapi(cls):
-        return __import__('ingresdbi')
+    if (int((sqlalchemy.__version__).split('.')[0]) >= 2):
```
Comment (clach04) [inline lib/ingres_sa_dialect/ingresdbi.py:21]: same comment on compare - maybe but shared logic in the base rather than each implementation file?
create tuple in base (as a global) and then compare is simple:

    (1, 2, 3) > ......

e.g.

    sqlalchemy_version_tuple - tuple(map(int, sqlalchemy.__version__.split('.')))  # assuming ther isn't one already we can make use of

File: lib/ingres_sa_dialect/pyodbc.py [hunk 4]
```diff
@@ -28,15 +27,26 @@ class Ingres_pyodbc(IngresDialect):
...
-    @classmethod
-    def dbapi(cls):
-        try:
-            driver = __import__(Ingres_pyodbc.driver)
-        except ModuleNotFoundError:
-            # fallback to pure Python version
-            Ingres_pyodbc.driver = 'pypyodbc'
-            driver = __import__(Ingres_pyodbc.driver)
-        return driver
+    if (int((sqlalchemy.__version__).split('.')[0]) >= 2):
```
Comment (clach04) [inline lib/ingres_sa_dialect/pyodbc.py:30]: ""

**Review by clach04 — CHANGES_REQUESTED** [review]
I love that this handles previous versions as well :-)

File: lib/ingres_sa_dialect/pyodbc.py [hunk 5]
```diff
@@ -28,15 +27,26 @@ class Ingres_pyodbc(IngresDialect):
     def __init__(self, **kwargs):
         IngresDialect.__init__(self, **kwargs)
 
-    @classmethod
-    def dbapi(cls):
```
Comment (clach04) [inline lib/ingres_sa_dialect/pyodbc.py:32]: Idea for next pass (i.e. no need to change for this commit). Push this logic into parent (base) class and use the pyodbc code where this is a variable NOT a string literal for lookup - to remove code duplication.

File: lib/ingres_sa_dialect/base.py (hunk 1 above)
Comment (hab6) [reply lib/ingres_sa_dialect/base.py]: Changed code to use tuple/map as suggested.

File: lib/ingres_sa_dialect/ingresdbi.py (hunk 3 above)
Comment (hab6) [reply lib/ingres_sa_dialect/ingresdbi.py]: Now using tuple variable defined in base.py.

File: lib/ingres_sa_dialect/ingresdbi.py (hunk 2 above)
Comment (hab6) [reply lib/ingres_sa_dialect/ingresdbi.py]: Rearranged import statements in accordance with https://peps.python.org/pep-0008/#imports

**Comment by hab6** [comment]
@clach04 I've implemented suggested changes and successfully retested with SQLAlchemy 1.4.46 and 2.0.3. Awaiting your subsequent re-review and approval in order to squash and merge.

File: lib/ingres_sa_dialect/ingresdbi.py [hunk 6]
```diff
@@ -17,9 +18,14 @@ class Ingres_ingresdbi(IngresDialect):
     def __init__(self, **kwargs):
         IngresDialect.__init__(self, **kwargs)
 
-    @classmethod
-    def dbapi(cls):
-        return __import__('ingresdbi')
+    if (sqlalchemy_version_tuple >= (2,0)):
+        @classmethod
+        def import_dbapi(cls):
```
Comment (clach04) [inline lib/ingres_sa_dialect/ingresdbi.py:23]: as per last comment, lets get a ticket opened for potential refactor here and push logic into base (if possible)

**Review by clach04 — APPROVED** [review]
lgtm! 👍 

Thanks for the refactoring.
